src/chainstate/stacks/miner.rs

```diff
@@ -1230,98 +1210,83 @@ impl StacksBlockBuilder {
-                    );
-                    warn!("{}", &msg);
-
-                    return Err(Error::InvalidStacksMicroblock(msg, mblock_header_hash));
-                }
-            };
             let num_mblocks = parent_microblocks.len();
             let last_mblock_hdr = parent_microblocks[num_mblocks - 1].header.clone();
             self.set_parent_microblock(&last_mblock_hdr.block_hash(), last_mblock_hdr.sequence);
+        };
 
-            let mut microblock_cost = tx.cost_so_far();
-            microblock_cost
-                .sub(&parent_block_cost)
-                .expect("BUG: block_cost + microblock_cost < block_cost");
-
-            // if we get here, then we need to reset the block-cost back to 0 because this begins the
-            // block defined by this miner.
-            tx.reset_cost(ExecutionCost::zero());
+        let parent_consensus_hash = self.chain_tip.consensus_hash.clone();
+        let parent_header_hash = self.header.parent_block.clone();
 
-            microblock_cost
-        };
+        let mainnet = chainstate.config().mainnet;
 
-        let t2 = get_epoch_time_ms();
+        let (chainstate_tx, clarity_instance) = chainstate.chainstate_tx_begin()?;
 
-        debug!(
-            "Miner {}: Finished applying {} parent microblocks in {}ms\n",
-            self.miner_id,
-            parent_microblocks.len(),
-            t2.saturating_sub(t1)
-        );
+        Ok((
+            chainstate_tx,
+            clarity_instance,
+            burn_tip,
+            burn_tip_height + 1,
+            parent_microblocks,
+            parent_consensus_hash,
+            parent_header_hash,
+            mainnet,
+        ))
+    }
 
-        StacksChainState::process_epoch_transition(&mut tx, burn_tip_height + 1)?;
-        StacksChainState::process_stacking_ops(&mut tx, stacking_burn_ops);
-        StacksChainState::process_transfer_ops(&mut tx, transfer_burn_ops);
+    /// Begin mining an epoch's transactions.
+    /// Returns an open ClarityTx for mining the block, as well as the ExecutionCost of any confirmed
+    ///  microblocks.
+    /// NOTE: even though we don't yet know the block hash, the Clarity VM ensures that a
+    /// transaction can't query information about the _current_ block (i.e. information that is not
+    /// yet known).
+    /// This function was separated from `pre_epoch_begin` because to return a ClarityTx object,
+    /// we need to pass ChainstateTx & ClarityInstance to `epoch_begin` directly instead of
+    /// passing the Chainstate object.
+    pub fn epoch_begin<'a>(
+        &mut self,
+        chainstate_tx: &'a mut ChainstateTx,
+        clarity_instance: &'a mut ClarityInstance,
+        burn_dbconn: &'a SortitionDBConn,
+        burn_tip: BurnchainHeaderHash,
+        burn_tip_height: u32,
+        parent_microblocks: Vec<StacksMicroblock>,
+        parent_consensus_hash: ConsensusHash,
+        parent_header_hash: BlockHeaderHash,
+        mainnet: bool,
```

Comment:
A lot of these arguments, such as the burnchain and parent block data, could be constructor arguments to `StacksBlockBuilder`.  Alternatively, they could be emitted entirely if we got rid of `pre_epoch_begin()` and just pulled the relevant data from `pre_epoch_begin()`'s code body.